File: RCVLinearLayoutManager/app/src/main/java/com/example/rcvlinearlayoutmanager/UserAdapter.java
package com.example.rcvlinearlayoutmanager;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class UserAdapter extends RecyclerView.Adapter<UserAdapter.UserViewHolder> {

    private Context context;
    private List<User> mListUser;

    public UserAdapter(Context context) {
        this.context = context;
    }

    public void setData(List<User> list) {
        this.mListUser = list;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_user, parent, false);

        return new UserViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        User user = mListUser.get(position);
        if (user == null) {
            return;
        }
        holder.tvKey.setText(user.getKey());
        holder.tvValue.setText(user.getValue());
    }

    @Override
    public int getItemCount() {
        if (mListUser != null) {
            return mListUser.size();
        }
        return 0;
    }

    public class UserViewHolder extends RecyclerView.ViewHolder {

        private TextView tvKey, tvValue;

        public UserViewHolder(@NonNull View itemView) {
            super(itemView);
            tvKey = itemView.findViewById(R.id.Key);
            tvValue = itemView.findViewById(R.id.Value);
        }
    }

}

File: LoadMoreKotlin/app/src/main/java/com/example/loadmorekotlin/UserAdapter.kt
package com.example.loadmorekotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import javax.inject.Inject

class UserAdapter @Inject constructor() : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        const val TYPE_ITEM = 1
        const val TYPE_LOADING = 2
    }

    private var listUser: MutableList<String> = mutableListOf()
    private var isLoading: Boolean = false

    fun setData(list: MutableList<String>) {
        this.listUser = list
        notifyDataSetChanged()
    }

    fun addFooterLoading() {
        isLoading = true
        listUser.add("")
    }

    fun removeLoading() {
        isLoading = false
        listUser.removeAt(listUser.size - 1)
        notifyItemRemoved(listUser.size - 1)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_ITEM -> {
                val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_user, parent, false)
                UserVH(view)
            }
            else -> {
                val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_loading, parent, false)
                LoadingVH(view)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder.itemViewType == TYPE_ITEM) {
            val user = listUser[position]
            val userVH: UserVH = holder as UserVH
            userVH.name.text = user + " " + (position + 1)
        }
    }

    override fun getItemCount(): Int {
        return listUser.size
    }

    override fun getItemViewType(position: Int): Int {
        if (position == listUser.size - 1 && isLoading) {
            return TYPE_LOADING
        }
        return TYPE_ITEM
    }

    class UserVH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val name: TextView = itemView.findViewById(R.id.tv_user)
    }

    class LoadingVH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val progress: ProgressBar = itemView.findViewById(R.id.progress_bar)
    }
}